p4include/psa.p4

```diff
@@ -25,403 +24,518 @@ limitations under the License.
+/// For any copies made of this packet at the end of Egress
+/// processing, truncate the payload to at most payload_bytes bytes in
+/// length.  A PSA implementation need not support truncation for
+/// recirculated packets.
+
+action egress_truncate(inout psa_egress_output_metadata_t meta,
+                       in PacketLength_t payload_bytes)
+{
+    meta.truncate = true;
+    meta.truncate_payload_bytes = payload_bytes;
+}
+// END:Action_egress_truncate
 
-/**
- * The PacketReplicationEngine extern represents the non-programmable
- * part of the PSA pipeline.
- *
- * Even though the PRE can not be programmed using P4, it can be
- * configured both directly using control plane APIs and by setting
- * intrinsic metadata. In this specification we opt to define the
- * operations available in the PRE as method invocations. A target
- * backend is responsible for mapping the PRE extern APIs to the
- * appropriate mechanisms for performing these operations in the
- * hardware.
- *
- * The PRE is instantiated by the architecture and a P4 program can
- * use it directly. It is an error to instantiate the PRE multiple
- * times.  The PRE is made available to the Ingress and Egress
- * programmable blocks using the same mechanism as packet_in.
- *
- * Note: some of these operations may not be implemented as primitive
- * operations on a certain target. However, All of the operations can
- * be implemented as a combination of other operations. Applications
- * that rely on non-primitive operations may incur significant
- * performance penalties, however, they should be functionally
- * correct.
- *
- * Alternatively, we can consider a design in which we split the PRE
- * (and the associated functionality) between an ingress PRE extern
- * and a egress BQE (Buffering and Queueing Engine) extern. That
- * decision will be finalized based on application needs.
- *
- * Semantics of behavior for multiple calls to PRE APIs
- *
- * - multiple calls to send_to_port -- the last call in the ingress
- *   pipeline sets the output port.
- *
- * - multiple calls to multicast -- the last in the ingress pipeline
- *   sets the multicast group
- *
- * - interleaving send_to_port and multicast -- the semantics of multicast
- *   is defined as below (https://github.com/p4lang/tutorials/issues/22):
- *     if (multicast_group != 0)
- *        multicast_to_group(multicast_group);
- *     else
- *        send_to_port(multicast_port);
- * From this, it follows that if there is a call that sets the
- * multicast_group, the packet will be multicast to the group that
- * last set the multicast group. Otherwise, the packet will be sent to
- * the port set by either send_to_port or multicast invocation.
- *
- * - any call to drop in the pipeline will cause the packet to drop
- *
- * - how do we drop clone packets?
- * ...
- *
- * \TODO: finalize the semantics of calling multiple of the PRE APIs
- *
- */
 extern PacketReplicationEngine {
+    PacketReplicationEngine();
```

Comment:
Maybe you want to delete the commend below.

Reply:
I think @cc10512 is still hoping the "No constructor.  PRE is instantiated by the architecture" comment becomes true, but I will delete the commented-out code for now.